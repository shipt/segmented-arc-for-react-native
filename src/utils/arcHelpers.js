export const polarToCartesian = (centerX, centerY, radius, angleInDegrees) => {
  const angleInRadians = ((angleInDegrees - 180) * Math.PI) / 180.0;

  return {
    x: centerX + radius * Math.cos(angleInRadians),
    y: centerY + radius * Math.sin(angleInRadians)
  };
};

export const drawArc = (x, y, radius, startAngle, endAngle, range = false) => {
  const start = polarToCartesian(x, y, radius, endAngle);
  const end = polarToCartesian(x, y, radius, startAngle);

  const largeArcFlag = endAngle - startAngle <= 180 ? '0' : '1';

  const sweepFlag = range ? 1 : 0;

  const d = ['M', start.x, start.y, 'A', radius, radius, 0, largeArcFlag, sweepFlag, end.x, end.y].join(' ');

  return d;
};
