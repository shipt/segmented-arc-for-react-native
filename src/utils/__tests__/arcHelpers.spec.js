import { polarToCartesian, drawArc } from '../arcHelpers';

describe('arc helpers', () => {
  it('converts polar coordinates to cartesian coordinates', () => {
    expect(polarToCartesian(100, 100, 100, 90)).toEqual({ x: 100, y: 0 });
  });

  it('returns d path for an arc', () => {
    expect(drawArc(152, 152, 104, 136.5, 158)).toMatchSnapshot();
  });

  it('returns d path for an arc - opposite angle needed for ratings range display', () => {
    expect(drawArc(152, 152, 104, 136.5, 158, true)).toMatchSnapshot();
  });
});
