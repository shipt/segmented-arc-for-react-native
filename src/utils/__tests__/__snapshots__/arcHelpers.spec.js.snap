// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`arc helpers returns d path for an arc - opposite angle needed for ratings range display 1`] = `"M 248.4271208749459 113.04091428474516 A 104 104 0 0 1 239.4680573482968 95.73865497772863 A 104 104 0 0 1 227.43893458527793 80.41112412784959"`;

exports[`arc helpers returns d path for an arc 1`] = `"M 248.4271208749459 113.04091428474516 A 104 104 0 0 0 239.4680573482968 95.73865497772863 A 104 104 0 0 0 227.43893458527793 80.41112412784959"`;
